library/std/src/sys/unix/ext/ucred.rs

```diff
@@ -0,0 +1,93 @@
+//! Unix peer credentials.
+
+// NOTE: Code in this file is heavily based on work done in PR 13 from the tokio-uds repository on
+//       GitHub.
+//
+//       For reference, the link is here: https://github.com/tokio-rs/tokio-uds/pull/13
+//       Credit to Martin Habovštiak (GitHub username Kixunil) and contributors for this work.
+
+use libc::{gid_t, uid_t};
+
+/// Credentials for a UNIX process for credentials passing.
+#[unstable(feature = "peer_credentials_unix_socket", issue = "42839", reason = "unstable")]
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct UCred {
+    pub uid: uid_t,
+    pub gid: gid_t,
+}
+
+#[cfg(any(target_os = "android", target_os = "linux"))]
+pub use self::impl_linux::peer_cred;
+
+#[cfg(any(
+    target_os = "dragonfly",
+    target_os = "freebsd",
+    target_os = "ios",
+    target_os = "macos",
+    target_os = "openbsd"
+))]
+pub use self::impl_bsd::peer_cred;
+
+#[cfg(any(target_os = "linux", target_os = "android"))]
+pub mod impl_linux {
+    use super::UCred;
+    use crate::mem::MaybeUninit;
+    use crate::os::unix::io::AsRawFd;
+    use crate::os::unix::net::UnixStream;
+    use crate::{io, mem};
+
+    pub fn peer_cred(socket: &UnixStream) -> io::Result<UCred> {
+        use libc::{c_void, ucred};
+
+        let ucred_size = mem::size_of::<ucred>();
+
+        // Trivial sanity checks.
+        assert!(mem::size_of::<u32>() <= mem::size_of::<usize>());
+        assert!(ucred_size <= u32::max_value() as usize);
+
+        let mut ucred_size = ucred_size as u32;
+
+        unsafe {
+            let mut ucred: ucred = MaybeUninit::uninit().assume_init();
```

Comment:
I don't think `assume_init` is correct here. I believe raw ptr should be obtained from `MaybeUninit`, then passed to `getsockopt` and only after it succeeds calling `assume_init`

Reply:
Done -- I opted against using `MemUninit` in the end, it doesn't seem necessary.